lib/jsonapi/resource.rb

```
@@ -302,6 +302,20 @@ def _replace_to_many_links(relationship_type, relationship_key_values, options)
         to_add = relationship_key_values - (relationship_key_values & existing)
         _create_to_many_links(relationship_type, to_add, {})
 
+        @reload_needed = true
+      elsif relationship.polymorphic?
+        relationship_resource_klass = self.class.resource_for(relationship_key_values[:type])
+        relationship_klass = relationship_resource_klass.model_hint.classify.safe_constantize
+        relation_name = relationship.relation_name(context: @context)
+
+        begin
+          related_records = relationship_klass.find(relationship_key_values[:ids])
+        rescue ActiveRecord::RecordNotFound
+          fail JSONAPI::Exceptions::RecordNotFound.new(relationship_key_values[:ids])
```

The reason to raise RecordNotFound instead of TypeMismatch is that type mismatch requires 2 conditions:
      - the id can't be used to find any record in current subclass (Car)
      - but in the meantime, it does exists for the superclass (Vehicle)
And this means we need to perform additional query to prove the second condition, which is kind of a waste.

---

I tend to agree it's kind of a waste, but from a JSON API perspective it's not known that the backing table is sharing a common set of IDs. So there really isn't a Car with id 2 so the system should raise an error.

---

Let's use `find_by_keys`, like https://github.com/cerebris/jsonapi-resources/blob/master/lib/jsonapi/resource.rb#L251. That way any overrides will get applied. This might be the case if people are doing permission checking for these related resources.